pandas/core/frame.py

```diff
@@ -4150,10 +4145,26 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
                 # tuples are always considered keys, never as list-likes
                 if col not in self:
                     missing.append(col)
-            elif (not isinstance(col, (ABCIndexClass, ABCSeries,
-                                       np.ndarray, list))
-                  or getattr(col, 'ndim', 1) > 1):
-                raise ValueError(err_msg)
+            elif isinstance(col, (ABCIndexClass, ABCSeries,
+                                  np.ndarray, list)):
+                # arrays are fine as long as they are one-dimensional
+                if getattr(col, 'ndim', 1) > 1:
+                    raise ValueError(err_msg)
+            else:
+                # everything else gets tried as a key; see GH 24969
+                try:
+                    hash(col) and self[col]
+                except TypeError:
+                    # for unhashable types
+                    tipo = type(col)
+                    raise TypeError(err_msg,
+                                    'Received column of type {}'.format(tipo))
+                except KeyError:
+                    # for hashable types that are not keys;
+                    # treat as ValueError, not missing key
+                    tipo = type(col)
+                    raise ValueError(err_msg,
```

Comment:
What is the reason for not treating it as a missing key?

Reply:
Because at that point we're dealing with an object we have absolutely no idea about, and assuming it is a key is worse (see the `iter`-KeyError) than assuming it isn't one.

Reply:
No, I mean assuming it is a *missing* key. If it cannot be a key, it certainly is a missing key. Anyway, this is exactly the same behaviour as what `__getitem__` is also doing (or what 0.23.4 was doing), so I don't see the reason to deviate from that